From 18d93df7fda7fc6ee5097b167eb555b8059ec49e Mon Sep 17 00:00:00 2001
From: Heiko Stuebner <heiko@sntech.de>
Date: Sat, 9 Dec 2017 13:37:11 +0100
Subject: [PATCH 090/134] lima: handle Mali450 pmu correctly

The powerdomain handling on Mali450 is slightly different in that
the PPs are not controllable separately and the L2 caches are also
turned on automatically with their domain.

As we're currently just turn on everything anyway, adapt the power-up
accordingly for mali450 as well.

Signed-off-by: Heiko Stuebner <heiko@sntech.de>
---
 drivers/gpu/drm/lima/lima_pmu.c | 29 ++++++++++++++++++++++++++---
 1 file changed, 26 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/lima/lima_pmu.c b/drivers/gpu/drm/lima/lima_pmu.c
index 9f705d8..2478325 100644
--- a/drivers/gpu/drm/lima/lima_pmu.c
+++ b/drivers/gpu/drm/lima/lima_pmu.c
@@ -5,6 +5,16 @@
 #define   LIMA_PMU_POWER_GP0_MASK          (1 << 0)
 #define   LIMA_PMU_POWER_L2_MASK           (1 << 1)
 #define   LIMA_PMU_POWER_PP_MASK(i)        (1 << (2 + i))
+
+/*
+ * On Mali450 each block automatically starts up its corresponding L2
+ * and the PPs are not fully independent controllable.
+ * Instead PP0, PP1-3 and PP4-7 can be turned on or off.
+ */
+#define   LIMA450_PMU_POWER_PP0_MASK       BIT(1)
+#define   LIMA450_PMU_POWER_PP13_MASK      BIT(2)
+#define   LIMA450_PMU_POWER_PP47_MASK      BIT(3)
+
 #define LIMA_PMU_STATUS                    0x08
 #define LIMA_PMU_INT_MASK                  0x0C
 #define LIMA_PMU_INT_RAWSTAT               0x10
@@ -48,9 +58,22 @@ int lima_pmu_init(struct lima_pmu *pmu)
 	stat = pmu_read(STATUS);
 
 	/* power up all ip */
-	mask = LIMA_PMU_POWER_GP0_MASK | LIMA_PMU_POWER_L2_MASK;
-	for (i = 0; i < dev->num_pp; i++)
-		mask |= LIMA_PMU_POWER_PP_MASK(i);
+	switch (dev->gpu_type) {
+	case GPU_MALI400:
+		mask = LIMA_PMU_POWER_GP0_MASK | LIMA_PMU_POWER_L2_MASK;
+		for (i = 0; i < dev->num_pp; i++)
+			mask |= LIMA_PMU_POWER_PP_MASK(i);
+		break;
+	case GPU_MALI450:
+		mask = LIMA_PMU_POWER_GP0_MASK | LIMA450_PMU_POWER_PP0_MASK;
+		if (dev->num_pp > 1)
+			mask |= LIMA450_PMU_POWER_PP13_MASK;
+		if (dev->num_pp > 4)
+			mask |= LIMA450_PMU_POWER_PP47_MASK;
+		break;
+	default:
+		return -ENODEV;
+	}
 
 	if (stat & mask) {
 		pmu_write(POWER_UP, stat & mask);
-- 
2.0.1

